test/tests_for_sdkmr/job_submission_parameters_test.py

```diff
@@ -207,24 +228,256 @@ def test_hash():
+    )
+    _job_sub_init_fail(
+        j,
+        n,
+        r,
+        u,
+        n,
+        n,
+        n,
+        ValueError("app_info cannot be a value that evaluates to false"),
+    )
+    _job_sub_init_fail(
+        j,
+        a,
+        n,
+        u,
+        n,
+        n,
+        n,
+        ValueError("job_reqs cannot be a value that evaluates to false"),
+    )
+    _job_sub_init_fail(
+        j,
+        a,
+        r,
+        n,
+        n,
+        n,
+        n,
+        ValueError("user_creds cannot be a value that evaluates to false"),
+    )
+    # the only way to get parent id to to fail is with a control char
+    _job_sub_init_fail(
+        j,
+        a,
+        r,
+        u,
+        "par\bent",
+        n,
+        n,
+        IncorrectParamsException("parent_job_id contains control characters"),
+    )
+    _job_sub_init_fail(
+        j, a, r, u, n, 0, n, IncorrectParamsException("wsid must be at least 1")
+    )
+    _job_sub_init_fail(
+        j,
+        a,
+        r,
+        u,
+        n,
+        n,
+        ["1/2/3", n],
+        IncorrectParamsException(
+            "source_ws_objects index 1, 'None', is not a valid Unique Permanent Address"
+        ),
+    )
+    _job_sub_init_fail(
+        j,
+        a,
+        r,
+        u,
+        n,
+        n,
+        ["1/2/3", "   \t  "],
+        IncorrectParamsException(
+            "source_ws_objects index 1, '   \t  ', is not a valid Unique Permanent Address"
+        ),
+    )
+    for o in ["1/2", "1/2/", "/1/2", "1/2/3/4", "x/2/3", "1/x/3", "1/2/x"]:
+        _job_sub_init_fail(
+            j,
+            a,
+            r,
+            u,
+            n,
+            n,
+            [o],
+            IncorrectParamsException(
+                f"source_ws_objects index 0, '{o}', is not a valid Unique Permanent Address"
+            ),
+        )
+
+
+def _job_sub_init_fail(jobid, appinfo, jobreq, usercred, parentid, wsid, wso, expected):
+    with raises(Exception) as got:
+        JobSubmissionParameters(jobid, appinfo, jobreq, usercred, parentid, wsid, wso)
+    assert_exception_correct(got.value, expected)
+
+
+def test_job_sub_equals():
+    j1 = "jobby job job"
+    j1a = "jobby job job"
+    j2 = "jobby job job JOB"
+    a1 = AppInfo("a.b", "a/x")
+    a1a = AppInfo("a.b", "a/x")
+    a2 = AppInfo("a.b", "a/y")
+    r1 = JobRequirements(6, 7, 4, "cligroup")
+    r1a = JobRequirements(6, 7, 4, "cligroup")
+    r2 = JobRequirements(6, 7, 4, "cligroup2")
+    u1 = UserCreds("user", "tokeytoken")
+    u1a = UserCreds("user", "tokeytoken")
+    u2 = UserCreds("user", "tokeytoken2")
+    p1 = "I'm so miserable and you just don't care"
+    p1a = "I'm so miserable and you just don't care"
+    p2 = "Oh do shut up Portia"
+    w1 = ["1/2/3"]
+    w1a = ["1/2/3"]
+    w2 = ["1/2/4"]
+
+    JSP = JobSubmissionParameters
+
+    assert JSP(j1, a1, r1, u1) == JSP(j1a, a1a, r1a, u1a)
```

Comment:
again, sort and save params/JSP objects to variables

Reply:
sort same question as above, save vars done